### ISSUE TITLE
`paddle.nn.functional.scaled_dot_product_attention`在输入为float32类型时，性能非常差

### ISSUE THREAD

**Issue by HydrogenSulfate** [PaddlePaddle/Paddle#76956]
### bug描述 Describe the Bug

`paddle.nn.functional.scaled_dot_product_attention`在输入为float32类型时，反向性能非常差，耗时为torch的17倍，导致模型无法正常训练

A100 40G卡上性能如下：

前向：paddle 18ms, torch 20ms
反向：paddle 1010ms, torch 58ms

测试代码：

``` py
import paddle
import torch
from contextlib import ContextDecorator
import time


class PDTimer(ContextDecorator):
    interval: float  # Time cost for code within Timer context

    def __init__(self, name: str = "Timer", auto_print: bool = True):
        super().__init__()
        self.name = name
        self.auto_print = auto_print

    def __enter__(self):
        paddle.device.synchronize()
        self.start_time = time.perf_counter()
        return self

    def __exit__(self, type, value, traceback):
        paddle.device.synchronize()
        self.end_time = time.perf_counter()
        self.interval = self.end_time - self.start_time
        if self.auto_print:
            print(f"{self.name}.time_cost = {self.interval * 1000:.2f} ms")

    def start(self, name: str = "Timer"):
        paddle.device.synchronize()
        self.start_time = time.perf_counter()

    def end(self):
        paddle.device.synchronize()
        self.end_time = time.perf_counter()
        self.interval = self.end_time - self.start_time
        if self.auto_print:
            print(f"{self.name}.time_cost = {self.interval:.2f} s")


class PTTimer(ContextDecorator):
    interval: float  # Time cost for code within Timer context

    def __init__(self, name: str = "Timer", auto_print: bool = True):
        super().__init__()
        self.name = name
        self.auto_print = auto_print

    def __enter__(self):
        torch.cuda.synchronize()
        self.start_time = time.perf_counter()
        return self

    def __exit__(self, type, value, traceback):
        torch.cuda.synchronize()
        self.end_time = time.perf_counter()
        self.interval = self.end_time - self.start_time
        if self.auto_print:
            print(f"{self.name}.time_cost = {self.interval * 1000:.2f} ms")

    def start(self, name: str = "Timer"):
        torch.cuda.synchronize()
        self.start_time = time.perf_counter()

    def end(self):
        torch.cuda.synchronize()
        self.end_time = time.perf_counter()
        self.interval = self.end_time - self.start_time
        if self.auto_print:
            print(f"{self.name}.time_cost = {self.interval:.2f} s")



def test_paddle_sdpa():
    q = paddle.randn([1, 8, 16384, 32], requires_grad=True)
    k = paddle.randn([1, 8, 16384, 32], requires_grad=True)
    v = paddle.randn([1, 8, 16384, 32], requires_grad=True)

    tq = torch.from_dlpack(q.detach()).requires_grad_(True)
    tk = torch.from_dlpack(k.detach()).requires_grad_(True)
    tv = torch.from_dlpack(v.detach()).requires_grad_(True)
    for i in range(5):
        with PDTimer("paddle sdpa forward"):
            out = paddle.nn.functional.scaled_dot_product_attention(
                q.transpose([0, 2, 1, 3]),
                k.transpose([0, 2, 1, 3]),
                v.transpose([0, 2, 1, 3]),
                attn_mask=None,
                dropout_p=0.0,
                # training=True,
                enable_gqa=False,
            ).transpose([0, 2, 1, 3])
            if i == 0:
                print(out.shape)
            dout = paddle.randn_like(out)
        with PDTimer("paddle sdpa backward"):
            ds = paddle.grad(out, [q, k, v], dout)

        tdout = torch.from_dlpack(dout)
        with PTTimer("torch sdpa forward"):
            tout = torch.nn.functional.scaled_dot_product_attention(
                tq,
                tk,
                tv,
                attn_mask=None,
                dropout_p=0.0,
            )
        if i == 0:
            print(tout.shape)
        with PTTimer("torch sdpa backward"):
            tds = torch.autograd.grad(tout, [tq, tk, tv], tdout)
        print("\n")
        torch.testing.assert_close(torch.from_dlpack(out.detach()), tout)
        torch.testing.assert_close(torch.from_dlpack(ds[0].detach()), tds[0])
        torch.testing.assert_close(torch.from_dlpack(ds[1].detach()), tds[1])
        torch.testing.assert_close(torch.from_dlpack(ds[2].detach()), tds[2])

test_paddle_sdpa()
```

**Comment by zhangyuqin1998** [comment]
感谢您的issue，我们正在排查

**Comment by PlumBlossomMaid** [comment]
根据[相关文档](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/functional/scaled_dot_product_attention_cn.html)中的描述，`scaled_dot_product_attention`中的输入仅支持`float16`和`bfloat16`的输入。
如果你想要整一个正常一点的`scaled_dot_product_attention`算法，可以参考：
```Python
def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0,
                is_causal=False, scale=None, enable_gqa=False) -> paddle.Tensor:
    L, S = query.shape[-2], key.shape[-2]
    scale_factor = 1 / math.sqrt(query.shape[-1]) if scale is None else scale
    attn_bias = paddle.zeros([L, S], dtype=query.dtype)
    if is_causal:
        assert attn_mask is None
        temp_mask = paddle.ones([L, S], dtype=paddle.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)
    
    if attn_mask is not None:
        if attn_mask.dtype == paddle.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias = attn_mask + attn_bias
    
    if enable_gqa:
        key = key.repeat_interleave(query.shape[-3]//key.shape[-3], -3)
        value = value.repeat_interleave(query.shape[-3]//value.shape[-3], -3)
    
    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = paddle.softmax(attn_weight, dim=-1)
    attn_weight = paddle.nn.functional.dropout(attn_weight, dropout_p, training=True)
    out = attn_weight @ value
    return out
```
确保使用最新版本paddle框架即可使用上述函数正确运行。
不过为什么`paddle.nn.functional.scaled_dot_product_attention`没有办法接受其他类型的输入呢？
闭上眼睛想都能猜出来肯定是paddle用C++算子实现了`paddle.nn.functional.scaled_dot_product_attention`，但是呢数据类型又没有支持全，就着急忙慌的发布了→_→

**Comment by HydrogenSulfate** [comment]
> 根据[相关文档](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/functional/scaled_dot_product_attention_cn.html)中的描述，`scaled_dot_product_attention`中的输入仅支持`float16`和`bfloat16`的输入。 如果你想要整一个正常一点的`scaled_dot_product_attention`算法，可以参考：
> 
> def scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0,
>                 is_causal=False, scale=None, enable_gqa=False) -> paddle.Tensor:
>     L, S = query.shape[-2], key.shape[-2]
>     scale_factor = 1 / math.sqrt(query.shape[-1]) if scale is None else scale
>     attn_bias = paddle.zeros([L, S], dtype=query.dtype)
>     if is_causal:
>         assert attn_mask is None
>         temp_mask = paddle.ones([L, S], dtype=paddle.bool).tril(diagonal=0)
>         attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
>         attn_bias.to(query.dtype)
>     
>     if attn_mask is not None:
>         if attn_mask.dtype == paddle.bool:
>             attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
>         else:
>             attn_bias = attn_mask + attn_bias
>     
>     if enable_gqa:
>         key = key.repeat_interleave(query.shape[-3]//key.shape[-3], -3)
>         value = value.repeat_interleave(query.shape[-3]//value.shape[-3], -3)
>     
>     attn_weight = query @ key.transpose(-2, -1) * scale_factor
>     attn_weight += attn_bias
>     attn_weight = paddle.softmax(attn_weight, dim=-1)
>     attn_weight = paddle.nn.functional.dropout(attn_weight, dropout_p, training=True)
>     out = attn_weight @ value
>     return out
> 确保使用最新版本paddle框架即可使用上述函数正确运行。 不过为什么`paddle.nn.functional.scaled_dot_product_attention`没有办法接受其他类型的输入呢？ 闭上眼睛想都能猜出来肯定是paddle用C++算子实现了`paddle.nn.functional.scaled_dot_product_attention`，但是呢数据类型又没有支持全，就着急忙慌的发布了→_→

谢谢回复，用自带的API还考虑到了显存的问题，在`q/k/v = [1, 32*32*16, 256]`的情况下，用类似于你的手写attn，训练显存会完全不够，而torch的sdpa即便是float32，也至少执行的是memory efficient实现

**Comment by PlumBlossomMaid** [comment]
emmmmmm如果是想要处理这种超大的Tensor的话，似乎只有两个解决方案，一个是把数据手动转换成float16，另一个是去找找其他库里面的实现，比如`PaddleNLP`，或许会有更加高效的实现。

**Comment by PlumBlossomMaid** [comment]
> A100 40G卡上性能如下：
> 
> 前向：paddle 18ms, torch 20ms
> 反向：paddle 1010ms, torch 58ms

不过这个测试结果让我发现了一个新的点：虽然[文档](https://www.paddlepaddle.org.cn/documentation/docs/zh/api/paddle/nn/functional/scaled_dot_product_attention_cn.html)里面说不能用fp32，但是这个算子的前向计算速度似乎被优化到了极致，或许这个算子可以暂时被安全的用于推理的场景。
